packages/compass-crud/src/components/bulk-update-dialog.tsx

```diff
@@ -0,0 +1,306 @@
+const codeEditorStyles = css({
+  background: 'transparent !important',
+  '& .cm-gutters': {
+    background: 'transparent !important',
+  },
+  '& .cm-editor': {
+    background: 'transparent !important',
+  },
+});
+
+const bannerContainerStyles = css({
+  // don't jump when an error appears
+  minHeight: spacing[4] * 2 + 2,
+  overflow: 'hidden',
+});
+
+const bannerStyles = css({
+  flex: 'none',
+  marginTop: spacing[2],
+  marginLeft: spacing[2],
+  marginRight: spacing[2],
+  textAlign: 'left',
+});
+
+const updatePreviewStyles = css({
+  display: 'flex',
+  flexDirection: 'column',
+  gap: spacing[3],
+});
+
+export type BulkUpdateDialogProps = {
+  isOpen: boolean;
+  ns: string;
+  filter: BSONObject;
+  count?: number;
+  updateText: string;
+  preview: UpdatePreview;
+  syntaxError?: Error & { loc?: { index: number } };
+  serverError?: Error;
+  closeBulkUpdateDialog: () => void;
+  updateBulkUpdatePreview: (updateText: string) => void;
+  runBulkUpdate: () => void;
+};
+
+type QueryLabelProps = {
+  tooltip: string;
+  label: string;
+};
+
+const queryLabelStyles = css({
+  display: 'flex',
+  gap: spacing[2],
+  alignItems: 'center',
+});
+
+const QueryLabel: React.FunctionComponent<QueryLabelProps> = ({
+  tooltip,
+  label,
+}) => {
+  return (
+    <div className={queryLabelStyles}>
+      <Label htmlFor="template-dropdown">{label}</Label>
+      <InfoSprinkle align="right">{tooltip}</InfoSprinkle>
+    </div>
+  );
+};
+
+export default function BulkUpdateDialog({
+  isOpen,
+  ns,
+  filter,
+  count,
+  updateText,
+  preview,
+  syntaxError,
+  serverError,
+  closeBulkUpdateDialog,
+  updateBulkUpdatePreview,
+  runBulkUpdate,
+}: BulkUpdateDialogProps) {
+  const darkMode = useDarkMode();
+
+  const [text, setText] = useState(updateText);
+
+  const previewDocuments = useMemo(() => {
+    return preview.changes.map(
+      (change) => new HadronDocument(change.after as Record<string, unknown>)
+    );
+  }, [preview]);
+
+  const onChangeText = (value: string) => {
+    setText(value);
+    updateBulkUpdatePreview(value);
+  };
+
+  const title =
+    count === undefined ? 'Update documents' : `Update documents (${count})`;
+
+  const annotations = useMemo<Annotation[]>(() => {
+    if (syntaxError?.loc?.index) {
+      return [
+        {
+          message: syntaxError.message,
+          severity: 'error',
+          from: syntaxError.loc.index,
+          to: syntaxError.loc.index,
+        },
+      ];
+    }
+
+    return [];
+  }, [syntaxError]);
+
+  return (
+    <FormModal
+      title={title}
+      subtitle={ns}
+      size="large"
+      open={isOpen}
+      onSubmit={runBulkUpdate}
+      onCancel={closeBulkUpdateDialog}
+      cancelButtonText="Close"
+      submitButtonText="Update documents"
+      submitDisabled={!!(syntaxError || serverError)}
+    >
+      <div className={columnsStyles}>
+        <div className={queryStyles}>
+          <div className={queryFieldStyles}>
+            <TextInput
```

Comment:
Do we want to switch the font to monospace here? That's what designs seems to expect

Reply:
I just copy/pasted from the bulk delete modal. We're doing them in parallel and reinventing every wheel twice. I'd rather have a ticket to just factor these out and clean it up afterwards.

Reply:
Got it, a ticket to clean this up afterwards sounds good 👍

Reply:
https://jira.mongodb.org/browse/COMPASS-7387